examples/cvector-generator/pca.hpp

```diff
@@ -28,288 +28,303 @@ static void print_debug_tensor(struct ggml_tensor * t, bool with_data = true) {
     printf(" ... ]\n");
 }
 
+// begin vanilla pca namespace
 namespace PCA {
 
 // input params for PCA computations
 struct pca_params {
-    int n_threads = 1;
-    int n_batch = 20; // number of iterations do to in one batch. larger the batch, more memory is used
+    int n_threads    = 1;
+    int n_batch      = 20; // number of iterations do to in one batch. larger the batch, more memory is used
     int n_iterations = 1000;
-    float tolerance = 1e-7;
-
-    // for debugging
-    int i_layer = 0;
-    int n_layers = 0;
+    float tolerance  = 1e-7;
 };
 
 // result from each iteration
 struct pca_result {
-    struct ggml_tensor * calculated_square = NULL;
-    std::vector<struct ggml_tensor *> eigenvectors;
-    std::vector<float> distances;
+    float * principal_component; // eigenvectors of the covariance matrix
+    float explained_variance;                 // eigenvalues of the covariance matrix
 };
 
-struct pca_model {
-    ggml_backend_t backend = NULL;
-    ggml_backend_buffer_t buffer;
-    struct ggml_context * ctx;      // context to compute graph on target device
-    struct ggml_context * ctx_host; // host context to store results
+static void compute_covariance(struct pca_params &pca_params,
+                        struct ggml_tensor * X,
+                        float * covariance,
+                        struct ggml_backend * backend) {
+
+    size_t ctx_size = 0;
+    ctx_size += 7 * X->ne[0] * X->ne[1] * ggml_type_size(GGML_TYPE_F32);
```

Comment:
This memory is not necessary, since the context is `no_alloc`, you only ened the tensor and graph overhead memory.